server/src/test/java/org/elasticsearch/common/util/CancellableSingleObjectCacheTests.java

```diff
@@ -222,18 +253,178 @@ protected String getKey(String s) {
+                    }, e -> {
+                        finishLatch.countDown();
+                        if (e instanceof TaskCancelledException) {
+                            assertTrue(cancel);
+                        } else {
+                            throw new AssertionError("unexpected", e);
+                        }
+                    });
+                });
+            }
+
+            for (int i = 0; i < 10; i++) {
+                threadPool.generic().execute(Objects.requireNonNull(queue.poll()));
+            }
+
+            startLatch.countDown();
+            assertTrue(finishLatch.await(10, TimeUnit.SECONDS));
+        } finally {
+            ThreadPool.terminate(threadPool, 10, TimeUnit.SECONDS);
+        }
+    }
+
+    public void testForegroundRefreshCanBeCancelled() throws InterruptedException {
+
+        final Runnable awaitBarrier = new Runnable() {
+            final CyclicBarrier barrier = new CyclicBarrier(2);
+
+            @Override
+            public void run() {
+                try {
+                    barrier.await(10, TimeUnit.SECONDS);
+                } catch (InterruptedException | BrokenBarrierException | TimeoutException e) {
+                    throw new AssertionError("unexpected", e);
+                }
+            }
+        };
+
+        final AtomicInteger successfulChecksRemaining = new AtomicInteger(between(2, 10));
+
+        final CancellableSingleObjectCache<String, String, Integer> testCache = new CancellableSingleObjectCache<>() {
+            @Override
+            protected void refresh(String s, Runnable ensureNotCancelled, ActionListener<Integer> listener, BooleanSupplier isFreshCheck) {
+                ActionListener.completeWith(listener, () -> {
+                    while (successfulChecksRemaining.get() > 0) {
+                        ensureNotCancelled.run();
+                    }
+                    awaitBarrier.run();
+                    awaitBarrier.run();
+                    ensureNotCancelled.run();
+                    if (s.equals("cancelled")) {
+                        throw new AssertionError("should have been cancelled");
+                    } else {
+                        return s.length();
+                    }
+                });
+            }
+
+            @Override
+            protected String getKey(String s) {
+                return s;
+            }
+        };
+
+        final TestFuture oldFuture = new TestFuture();
+        final TestFuture newFuture = new TestFuture();
+
+        final Thread cancelledThread = new Thread(() -> {
+            testCache.get("cancelled", () -> successfulChecksRemaining.getAndDecrement() <= 0, oldFuture);
+            awaitBarrier.run();
+            awaitBarrier.run();
+        }, "cancelled-thread");
+
+        cancelledThread.start();
+        awaitBarrier.run();
+        testCache.get("successful", () -> false, newFuture);
+        awaitBarrier.run();
```

Comment:
AFAICS, this matches the last `awaitBarrier` in the `cancelledThread` and I think we could skip the extra `awaitBarrier` in the `cancelledThread` and here - and just rely on the `cancelledThread.join`.

Which makes me wonder if I misunderstood the intention or failed to see the idea somehow?

Reply:
I'm using a pair of barriers to interleave execution on distinct threads which I find to be a handy test pattern but I guess it's not that obvious in this simple setting. The general idea is that one of the threads starts with a single `await()` to offset its execution relative to the other thread, and then every time either thread wants to yield to the other one it calls `await()` twice. Finally the main thread finishes with a single `await()` to accept execution back from the other thread. You're right however that the terminal odd `await()` could happen on the other thread.